### PR TITLE
Target_ajustements: initial commit

### LINKED ISSUE #84: Use the simple cpp generator model when targetP < pMin or targetP > pMax
We have seen cases in examples where in the iidm targetP (or p) is lower than pMin or higher than pMax.
In that case we should not implement a generator model in the .dyd file but do nothing and let dynawo use a simple injector cpp model.
@marcochiaramello what do you think ?

### PR DESCRIPTION
**Pull request by PerninGuillaume**
closes #84 #87 
Signed-off-by: Guillaume Pernin <guillaume.pernin@rte-france.com>

### CODE REVIEW

File: sources/Inputs/src/NetworkManager.cpp [hunk 1]
```diff
@@ -98,10 +98,11 @@ NetworkManager::buildTree() {
       // node should exist at this point
       assert(nodes_.count(nodeid));
 #endif
-      if ((*it_g)->isVoltageRegulationOn()) {
+      auto targetP = (*it_g)->getTargetP();
+      if ((*it_g)->isVoltageRegulationOn() && targetP >= (*it_g)->getPMin() && targetP <= (*it_g)->getPMax()) {
         // We don't use dynamic models for generators with voltage regulation disabled
```
Comment (cossartque) [inline sources/Inputs/src/NetworkManager.cpp:103]: We don't use dynamic models for generators with voltage regulation disabled and an active power reference outside the generator's PQ diagram

File: sources/Inputs/src/NetworkManager.cpp [hunk 2]
```diff
@@ -98,10 +98,11 @@ NetworkManager::buildTree() {
       // node should exist at this point
       assert(nodes_.count(nodeid));
 #endif
-      if ((*it_g)->isVoltageRegulationOn()) {
+      auto targetP = (*it_g)->getTargetP();
+      if ((*it_g)->isVoltageRegulationOn() && targetP >= (*it_g)->getPMin() && targetP <= (*it_g)->getPMax()) {
```
Comment (lecourtoisflo) [inline sources/Inputs/src/NetworkManager.cpp:102]: Use dynawo API isDoubleEqual for equality comparaison

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: tests/outputs/TestPar.cpp [hunk 3]
```diff
@@ -35,31 +35,34 @@ TEST(TestPar, write) {
...
                                                                      {
                                                                          GeneratorDefinition::ReactiveCurvePoint(1., 11., 110.),
                                                                          GeneratorDefinition::ReactiveCurvePoint(2., 22., 220.),
                                                                          GeneratorDefinition::ReactiveCurvePoint(3., 33., 330.),
                                                                          GeneratorDefinition::ReactiveCurvePoint(4., 44., 440.),
                                                                      },
-                                                                     4., 40., 44., 440.)};
+                                                                     4., 40., 44., 440., 100.),
+                                                 GeneratorDefinition("G4", GeneratorDefinition::ModelType::DIAGRAM_PQ_SIGNALN, "04",
+                                                                     {},
```
Comment (cossartque) [inline tests/outputs/TestPar.cpp:64]: Can you also test the case of a generator with targetP = 0 and without a diagram ?
This way we could see that for generators without diagrams we now have two sets of parameters : one with KGover=1 and one with KGover=0 and that in the .dyd we have the proper par set for each generator

**Review by cossartque — CHANGES_REQUESTED** [review]

**Review by lecourtoisflo — APPROVED** [review]

File: tests/outputs/TestPar.cpp (hunk 3 above)
Comment (PerninGuillaume) [reply tests/outputs/TestPar.cpp:64]: Hum, I did not think of that, that means that for every constant set in the par (such as impSignalNGenerator or signalNGenerator) we will duplicate it with the only thing changing being the "generator_KGover" ?

Comment (cossartque) [reply tests/outputs/TestPar.cpp:64]: That's it yes !
We would have impSignalNGenerator, signalNGenerator (with KGover = 1) and impSignalNGeneratorFixedP, signalNGeneratorFixedP (with KGover = 0) (any other names welcome !)

Comment (cossartque) [reply tests/outputs/TestPar.cpp:64]: And in the .dyd each generator could refer to each of these four sets depending on its situation

File: sources/Outputs/src/Par.cpp [hunk 4]
```diff
@@ -131,12 +136,23 @@ Par::writeConstantSets(unsigned int nb_generators, dfl::inputs::Configuration::A
...
-  updateSignalNGenerator(set, activePowerCompensation);
+  updateSignalNGenerator(set, activePowerCompensation, false);
+  ret.push_back(set);
+
+  // Signal N generator with fixed P
+  set = boost::shared_ptr<parameters::ParametersSet>(new parameters::ParametersSet(constants::signalNGeneratorFixedPParId));
+  updateSignalNGenerator(set, activePowerCompensation, true);
   ret.push_back(set);
 
   // Signal N generator with impendance
```
Comment (cossartque) [inline sources/Outputs/src/Par.cpp:147]: impedance instead of impendance :)

File: sources/Outputs/src/Par.cpp [hunk 5]
```diff
@@ -131,12 +136,23 @@ Par::writeConstantSets(unsigned int nb_generators, dfl::inputs::Configuration::A
...
   set = boost::shared_ptr<parameters::ParametersSet>(new parameters::ParametersSet(constants::impSignalNGeneratorParId));
-  updateSignalNGenerator(set, activePowerCompensation);
+  updateSignalNGenerator(set, activePowerCompensation, false);
+  updateCouplingParameters(set);
+  ret.push_back(set);
+
+  // Signal N generator with impendance with fixed P
```
Comment (cossartque) [inline sources/Outputs/src/Par.cpp:153]: same

File: sources/Outputs/src/Par.cpp (hunk 4 above)
Comment (PerninGuillaume) [reply sources/Outputs/src/Par.cpp:147]: That's not mine but I will fix it haha

**Review by cossartque — APPROVED** [review]

**Review by rosiereflo — CHANGES_REQUESTED** [review]

**Review by rosiereflo — APPROVED** [review]

**Review by marcochiaramello — APPROVED** [review]

File: sources/Outputs/include/Par.h [hunk 6]
```diff
@@ -95,9 +95,10 @@ class Par {
    *
    * @param set the parameter set to update
    * @param activePowerCompensation the type of active power compensation
+   * @param fixedP boolean to determines if the set represents a generator with a targetP equal to 0
```
Comment (lecourtoisflo) [inline sources/Outputs/include/Par.h:98]: typo: determine

File: sources/Outputs/src/Par.cpp [hunk 7]
```diff
@@ -194,8 +208,8 @@ Par::writeGenerator(const algo::GeneratorDefinition& def, const std::string& bas
   std::string hashIdStr = std::to_string(hashId);
   auto set = boost::shared_ptr<parameters::ParametersSet>(new parameters::ParametersSet(hashIdStr));
   //  Use the hash id in exported files to prevent use of non-ascii characters
-
-  set->addParameter(helper::buildParameter("generator_KGover", 1.));
+  double value = (DYN::doubleIsZero(def.targetP)) ? 0. : 1.;
```
Comment (lecourtoisflo) [inline sources/Outputs/src/Par.cpp:211]: These 2 values "0" and "1" could be factorized

File: tests/algo/TestAlgo.cpp [hunk 8]
```diff
@@ -168,25 +168,25 @@ TEST(Generators, base) {
...
+      dfl::algo::GeneratorDefinition("02", dfl::algo::GeneratorDefinition::ModelType::SIGNALN, "2", points, -2, 2, -2, 2, 0),
+      dfl::algo::GeneratorDefinition("05", dfl::algo::GeneratorDefinition::ModelType::SIGNALN, "4", points, -5, 5, -5, 5, 0),
   };
 
   dfl::algo::GeneratorDefinitionAlgorithm::Generators expected_gens_finite = {
-      dfl::algo::GeneratorDefinition("00", dfl::algo::GeneratorDefinition::ModelType::WITH_IMPEDANCE_DIAGRAM_PQ_SIGNALN, "0", points0, 0, 0, 0, 0),
-      dfl::algo::GeneratorDefinition("01", dfl::algo::GeneratorDefinition::ModelType::WITH_IMPEDANCE_DIAGRAM_PQ_SIGNALN, "0", points, -1, 1, -1, 1),
-      dfl::algo::GeneratorDefinition("02", dfl::algo::GeneratorDefinition::ModelType::DIAGRAM_PQ_SIGNALN, "2", points, -2, 2, -2, 2),
-      dfl::algo::GeneratorDefinition("05", dfl::algo::GeneratorDefinition::ModelType::DIAGRAM_PQ_SIGNALN, "4", points, -5, 5, -5, 5),
+      dfl::algo::GeneratorDefinition("00", dfl::algo::GeneratorDefinition::ModelType::WITH_IMPEDANCE_DIAGRAM_PQ_SIGNALN, "0", points0, 0, 0, 0, 0, 0),
```
Comment (lecourtoisflo) [inline tests/algo/TestAlgo.cpp:178]: Maybe add cases with targetP != 0, unless this parameter has no effect on algorithms

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/src/Par.cpp (hunk 7 above)
Comment (PerninGuillaume) [reply sources/Outputs/src/Par.cpp:211]: How would you do that ?

Comment (lecourtoisflo) [reply sources/Outputs/src/Par.cpp:211]: by defining them as constants of the class

File: tests/algo/TestAlgo.cpp (hunk 8 above)
Comment (PerninGuillaume) [reply tests/algo/TestAlgo.cpp:178]: The targetP has only an incidence on the Par.cpp and Dyd.cpp, so I don't think it is necessary

File: sources/Outputs/src/Par.cpp (hunk 7 above)
Comment (PerninGuillaume) [reply sources/Outputs/src/Par.cpp:211]: It is not very clear to me what you mean 
Do you want to add two variables in the class like double zero = 0.; double one = 1.; ?

**Review by lecourtoisflo — APPROVED** [review]

**Comment by rosiereflo** [comment]
is there a dependency with a dynawo commit or can we merge this right now?

**Pull request merged**